File: ProcedureStyle/container.cpp
// container.cpp - содержит функции обработки контейнера

#include "container.h"
#include "shape.h"

// Инициализация контейнера
void init(container& c) {
    c.len = 0;
}

// Очистка контейнера от элементов (освобождение памяти)
void clear(container& c) {
    for (int i = 0; i < c.len; i++) {
        delete c.cont[i];
    }
    c.len = 0;
}

// Ввод содержимого контейнера из указанного потока
void in(container &c, ifstream &ifst) {
    while(!ifst.eof()) {
        c.cont[c.len] = in(ifst);
        if(c.cont[c.len] != 0) {
            c.len++;
        }
    }
}

// Случайный ввод содержимого контейнера
void in_rand(container& c, int size) {
    while (c.len < size) {
        if ((c.cont[c.len] = in_rand()) != nullptr) {
            c.len++;
        }
    }
}

// Вывод содержимого контейнера в указанный поток
void out(container& c, ofstream& ofst) {
    ofst << "Container contains " << c.len << " elements. " << endl;
    for (int i = 0; i < c.len; i++) {
        ofst << i << ": ";
        out(*(c.cont[i]), ofst);
    }
}

// Сортировка фигур в контейнере по убыванию
void straightsel_sort(container& c) {
    int min;
    shape temp;
    for (int i = 0; i < c.len - 1; i++)
    {
        min = i;
        for (int j = i + 1; j < c.len; j++)
        {
            if (surf_sq(*c.cont[j]) < surf_sq(*c.cont[min]))
                min = j;
        }
        temp = *c.cont[i];
        *c.cont[i] = *c.cont[min];
        *c.cont[min] = temp;
    }
}

File: DynamicTyping/container.py
# container.ру контейнер для фигур
from random import randrange
from parallelepiped import parallelepiped
from regtet import regtet
from sphere import sphere


class container:
    # Контейнер на основе одномерного массива
    def __init__(self):
        self.store = []

    # Ввод содержимого контейнера из массива
    def read_string_arr(self, str_arr):
        i = 0  # Индекс, задающий текущую строку в массиве
        figNum = 0
        while i < len(str_arr):
            str = str_arr[i]
            print(str)
            key = int(str)
            if key == 1:
                i += 1
                shape = sphere()
                i = shape.read_string_arr(str_arr, i)
            elif key == 2:
                i += 1
                shape = parallelepiped()
                i = shape.read_string_arr(str_arr, i)
            elif key == 3:
                i += 1
                shape = regtet()
                i = shape.read_string_arr(str_arr, i)
            else:
                return figNum
            if i == 0:
                return figNum
            figNum += 1
            self.store.append(shape)
        return figNum

    # Случайный ввод содержимого контейнера
    def in_rand(self, count):
        for i in range(count):
            key = randrange(3) + 1
            if key == 1:
                rand_shape = sphere()
                rand_shape.in_rand()
                self.store.append(rand_shape)
            elif key == 2:
                rand_shape = parallelepiped()
                rand_shape.in_rand()
                self.store.append(rand_shape)
            elif key == 3:
                rand_shape = regtet()
                rand_shape.in_rand()
                self.store.append(rand_shape)
        pass

    # Вывод содержимого контейнера в консоль
    def print(self, sorted):
        print("Sorted container contains" if sorted else "Container contains", len(self.store), "elements:")
        for shape in self.store:
            shape.print()
        pass

    # Вывод содержимого контейнера в указанный поток
    def write(self, ostream, sorted):
        if (sorted):
            ostream.write("Container contains {} elements:\n".format(len(self.store)))
        else:
            ostream.write("Sorted container contains {} elements:\n".format(len(self.store)))
        for shape in self.store:
            shape.write(ostream)
            ostream.write("\n")
        pass

    # Сортировка фигур в контейнере по возрастанию
    def straightsel_sort(self):
        for i in range(len(self.store)):
            min_idx = i
            for j in range(i + 1, len(self.store)):
                if self.store[min_idx].surf_sq() > self.store[j].surf_sq():
                    min_idx = j
            self.store[i], self.store[min_idx] = self.store[min_idx], self.store[i]
        pass
